packages/firestore/src/api/credentials.ts

```diff
@@ -294,19 +297,39 @@ export class FirstPartyToken implements Token {
   user = User.FIRST_PARTY;
 
   constructor(
-    private gapi: Gapi,
-    private sessionIndex: string,
-    private iamToken: string | null
+    private readonly gapi: Gapi,
+    private readonly sessionIndex: string,
+    private readonly iamToken: string | null,
+    private readonly authTokenFactory: AuthTokenFactory | null
   ) {}
 
+  /** Gets an authorization token, using a provided factory function, or falling back to First Party GAPI. */
+  private getAuthToken(): string | null {
+    if (this.authTokenFactory) {
+      return this.authTokenFactory();
+    } else {
+      // Make sure this really is a Gapi client.
+      hardAssert(
+        !!(
+          typeof this.gapi === 'object' &&
+          this.gapi !== null &&
+          this.gapi['auth'] &&
+          this.gapi['auth']['getAuthHeaderValueForFirstParty']
+        ),
+        'unexpected gapi interface'
+      );
+      return this.gapi['auth']['getAuthHeaderValueForFirstParty']([]);
```

Comment:
Could we pass this block in as the default value for authTokenFactory()? We can then just invoke this.authTokenFactory for all callsites.